### PR TITLE
Be the Switch 🌟

### PR DESCRIPTION
**Pull request by daviddias**

### CODE REVIEW

**Comment by daviddias** [comment]
This PR revealed some issues that happen just on Linux, investigating.

```
  1) Switch (everything all together) close a muxer emits event:

     Uncaught AssertionError: expected true to not exist

      at conn.getPeerInfo (test/swarm-muxing.node.js:187:28)

      at node_modules/async/internal/once.js:12:16

      at next (node_modules/async/waterfall.js:21:29)

      at node_modules/async/internal/onlyOnce.js:12:16

      at f (node_modules/once/once.js:25:25)

      at select (node_modules/multistream-select/src/dialer/index.js:38:16)

      at pullLP.decodeFromReader (node_modules/multistream-select/src/select.js:20:14)

      at Object.reader.read [as cb] (node_modules/pull-length-prefixed/src/decode.js:88:16)

      at drain (node_modules/pull-reader/index.js:42:23)

      at node_modules/pull-reader/index.js:59:18

      at node_modules/pull-reader/index.js:20:7

      at drain (node_modules/stream-to-pull-stream/index.js:141:18)

      at Stream.<anonymous> (node_modules/stream-to-pull-stream/index.js:162:5)

      at node_modules/spdy-transport/lib/spdy-transport/stream.js:654:10

      at _combinedTickCallback (internal/process/next_tick.js:131:7)

      at process._tickCallback (internal/process/next_tick.js:180:9)


  2) Switch (everything all together) "after all" hook:

     Uncaught TypeError: Cannot set property 'state' of undefined

      at node_modules/async/internal/parallel.js:39:9

      at node_modules/async/internal/once.js:12:16

      at iteratorCallback (node_modules/async/eachOf.js:60:13)

      at node_modules/async/internal/onlyOnce.js:12:16

      at node_modules/async/internal/parallel.js:36:13

      at node_modules/async/internal/parallel.js:39:9

      at node_modules/async/internal/once.js:12:16

      at replenish (node_modules/async/internal/eachOfLimit.js:59:25)

      at iterateeCallback (node_modules/async/internal/eachOfLimit.js:49:17)

      at node_modules/async/internal/onlyOnce.js:12:16

      at node_modules/async/internal/parallel.js:36:13

      at node_modules/async/internal/once.js:12:16

      at iterateeCallback (node_modules/async/internal/eachOfLimit.js:47:24)

      at node_modules/async/internal/onlyOnce.js:12:16

      at node_modules/async/internal/once.js:12:16

      at Server.iteratorCallback (node_modules/async/eachOf.js:60:13)

      at Server.<anonymous> (node_modules/async/internal/onlyOnce.js:12:16)

      at emitCloseNT (net.js:1655:8)

      at _combinedTickCallback (internal/process/next_tick.js:135:11)

      at process._tickCallback (internal/process/next_tick.js:180:9)


  3) circuit listed on the transports map:

     AssertionError: expected undefined to exist

      at Context.it (test/circuit-relay.node.js:66:45)
```

**Comment by daviddias** [comment]
@dryajov any clue on the Circuit test fail?

**Comment by daviddias** [comment]
Got it! 😃

image

**Comment by daviddias** [comment]
@pgte @dryajov mind reviewing?

File: .aegir.js [hunk 1]
```diff
@@ -2,98 +2,87 @@
...
-        swarmB.connection.addStreamMuxer(spdy)
-        swarmB.connection.reuse()
-        swarmB.listen(cb)
-        swarmB.handle('/echo/1.0.0', echo)
-      })
+      peerA.multiaddrs.add(maA)
+      switchA = new Switch(peerA, new PeerBook())
+
+      switchA.transport.add('ws', new WebSockets())
+      switchA.transport.listen('ws', {}, echo, cb)
```
Comment (dryajov) [inline .aegir.js:42]: this is probably correct, but just to bring up the question, why no stream muxer here?

Comment (daviddias) [reply .aegir.js:42]: This one is for the transport only tests

**Comment by daviddias** [comment]
@VictorBjelkholm @dryajov related to https://github.com/ipfs/interop/issues/2, see how this module was refactored and one of the steps were dedup of tests. My point here is that the door is open for improvement, it was never locked :)

**Pull request merged**